# ADVENTUREWORKS  
## DIMPRODUCT

In [4]:
import dotenv
import os
from sqlalchemy import create_engine
import pandas as pd

username = os.getenv("username")
password = os.getenv("password")
host = os.getenv("host")
dbname = os.getenv("dbname")

dotenv.load_dotenv(override=True)
conn_string = "mysql+pymysql://" + username + ":" + password + "@" + host + "/" + dbname
db_engine = create_engine(conn_string)

In [6]:
query = "SELECT * FROM dimproduct"
df = pd.read_sql(sql=query, con=db_engine)
df

ProductKey ProductAlternateKey  ProductSubcategoryKey  \
0             1             AR-5381                    NaN   
1             2             BA-8327                    NaN   
2             3             BE-2349                    NaN   
3             4             BE-2908                    NaN   
4             5             BL-2036                    NaN   
..          ...                 ...                    ...   
601         602             BB-8107                    5.0   
602         603             BB-9108                    5.0   
603         604          BK-R19B-44                    2.0   
604         605          BK-R19B-48                    2.0   
605         606          BK-R19B-52                    2.0   

    WeightUnitMeasureCode SizeUnitMeasureCode     EnglishProductName  \
0                    None                None        Adjustable Race   
1                    None                None           Bearing Ball   
2                    None                None        BB Ball Bearing   
3                    None                None  Headset Ball Bearings   
4                    None                None                  Blade   
..                    ...                 ...                    ...   
601                     G                None      ML Bottom Bracket   
602                     G                None      HL Bottom Bracket   
603                    LB                  CM     Road-750 Black, 44   
604                    LB                  CM     Road-750 Black, 48   
605                    LB                  CM     Road-750 Black, 52   

            SpanishProductName           FrenchProductName  StandardCost  \
0                                                                    NaN   
1                                                                    NaN   
2                                                                    NaN   
3                                                                    NaN   
4                                                                    NaN   
..                         ...                         ...           ...   
601         Eje de pedalier GM          Axe de pédalier ML         44.95   
602         Eje de pedalier GA          Axe de pédalier HL         53.94   
603  Carretera: 750, negra, 44  Vélo de route 750 noir, 44        343.65   
604  Carretera: 750, negra, 48  Vélo de route 750 noir, 48        343.65   
605  Carretera: 750, negra, 52  Vélo de route 750 noir, 52        343.65   

     FinishedGoodsFlag  ...                 ChineseDescription  \
0                    0  ...                               None   
1                    0  ...                               None   
2                    0  ...                               None   
3                    0  ...                               None   
4                    0  ...                               None   
..                 ...  ...                                ...   
601                  1  ...                      铝合金车圈；大直径脚蹬轴。   
602                  1  ...                         铝合金车圈和空心轴。   
603                  1  ...  入门级成人自行车；确保越野旅行或公路骑乘的舒适。快拆式车毂和轮缘。   
604                  1  ...  入门级成人自行车；确保越野旅行或公路骑乘的舒适。快拆式车毂和轮缘。   
605                  1  ...  入门级成人自行车；确保越野旅行或公路骑乘的舒适。快拆式车毂和轮缘。   

                                     ArabicDescription  \
0                                                 None   
1                                                 None   
2                                                 None   
3                                                 None   
4                                                 None   
..                                                 ...   
601  فناجين من سبيكة الألومنيوم؛ ذات محور دوران كبي...   
602             فناجين من سبيكة الألومنيوم ومحور أجوف.   
603  إنها دراجة مناسبة للمبتدئين من البالغين؛ فهي ت...   
604  إنها دراجة مناسبة للمبتدئين من البالغين؛ فهي ت...   
605  إنها دراجة مناسبة للمبتدئين من البالغين؛ فهي ت

In [8]:
df.DealerPrice = df.DealerPrice.round(2)

In [10]:
df.loc[:,["ProductKey", "ProductAlternateKey", "DealerPrice"]].sample(20)

ProductKey ProductAlternateKey  DealerPrice
4             5             BL-2036          NaN
463         464           GL-H102-M        14.13
25           26             FW-3800          NaN
50           51             HJ-7162          NaN
53           54             HN-1024          NaN
144         145             LW-1400          NaN
17           18             EC-R098          NaN
215         216             HL-U509        20.19
516         517             SE-M940        31.58
520         521             SE-T312        16.27
325         326          BK-R50R-44       419.46
234         235           LJ-0192-X        28.84
446         447             LO-C100        15.00
593         594          BK-M18S-48       338.99
67           68             HN-5400          NaN
147         148             LW-4000          NaN
61           62             HN-3416          NaN
192         193             SA-T467          NaN
381         382          BK-R64Y-38       672.29
272         273          FR-R38R-62       202.33

In [12]:
df.DealerPrice = df.DealerPrice.round()

In [14]:
df.loc[:,["ProductKey", "ProductAlternateKey", "DealerPrice"]].sample(20)

ProductKey ProductAlternateKey  DealerPrice
135         136             LN-7162          NaN
442         443          FR-R92B-52        781.0
350         351          BK-M82B-48       2025.0
63           64             HN-3824          NaN
50           51             HJ-7162          NaN
400         401             HB-M918         66.0
215         216             HL-U509         20.0
429         430          FR-R72Y-40        357.0
21           22             FW-1000          NaN
467         468           GL-F110-S         23.0
586         587          BK-M38S-38        462.0
286         287          FR-R38B-52        202.0
472         473           VE-C304-L         38.0
3             4             BE-2908          NaN
428         429          FR-R72Y-40        324.0
362         363          BK-M68B-46       1377.0
371         372          BK-R89R-58       1466.0
533         534          FR-M21B-52        150.0
29           30             FW-9160          NaN
377         378          BK-R89B-52       1466.0

In [16]:
df.DealerPrice = df.DealerPrice.clip(upper = 100, lower = 0)

In [20]:
df.loc[:,["ProductKey", "ProductAlternateKey", "DealerPrice"]].sample(20)

ProductKey ProductAlternateKey  DealerPrice
405         406             HB-R720         37.0
451         452             LT-H903         27.0
115         116             LJ-9161          NaN
323         324          BK-R50R-62        100.0
67           68             HN-5400          NaN
407         408             HB-R956         72.0
275         276          FR-R72R-52        100.0
500         501             RD-2308         73.0
162         163             MS-1981          NaN
268         269          FR-R38R-58        100.0
526         527          FR-M21S-52        100.0
386         387          BK-R64Y-44        100.0
529         530             TT-T092          3.0
370         371          BK-R89R-58        100.0
226         227           LJ-0192-S         29.0
19           20             FE-3760          NaN
195         196             SD-2342          NaN
259         260          FR-R38B-62        100.0
338         339          BK-R50B-44        100.0
98           99             LI-7160          NaN

# DATAFRAME GUADAGNI MENSILI DIVERSE ANNATE

In [22]:
import numpy as np

In [24]:
years = 5
a = pd.DataFrame({"Mese": list("GFMAMGLASOND"*years),
                            "Anno": np.repeat(list(range(years)), 12),
                            "Guadagni": np.random.randint(800, 5000, 12*years)})

In [57]:
a

Mese  Anno  Guadagni
0     G     0      2382
1     F     0      2673
2     M     0      4956
3     A     0      4878
4     M     0      2899
5     G     0      3825
6     L     0      1026
7     A     0      1622
8     S     0      1578
9     O     0      3743
10    N     0       938
11    D     0      1773
12    G     1      2285
13    F     1       900
14    M     1      1196
15    A     1      2841
16    M     1      3027
17    G     1      2517
18    L     1       980
19    A     1      3225
20    S     1      1925
21    O     1      2873
22    N     1      4423
23    D     1      2223
24    G     2      2730
25    F     2      2707
26    M     2      4528
27    A     2      4930
28    M     2      2133
29    G     2      3960
30    L     2      1716
31    A     2      3777
32    S     2       959
33    O     2      3942
34    N     2      3065
35    D     2      1889
36    G     3      2832
37    F     3      3561
38    M     3      4076
39    A     3      1078
40    M     3      2185
41    G     3      1287
42    L     3      4345
43    A     3      2533
44    S     3      3910
45    O     3      4040
46    N     3      3286
47    D     3      2848
48    G     4      4967
49    F     4      4274
50    M     4      4769
51    A     4      1435
52    M     4      1740
53    G     4      2357
54    L     4      4397
55    A     4      3469
56    S     4      2279
57    O     4      4979
58    N     4      3523
59    D     4      3501

In [26]:
a.Mese = list( ["GENNAIO","FEBBRAIO","MARZO","APRILE","MAGGIO","GIUGNO",
                    "LUGLIO","AGOSTO","SETTEMBRE","OTTOBRE","NOVEMBRE","DICEMBRE"]*years)

In [28]:
a.Mese

0       GENNAIO
1      FEBBRAIO
2         MARZO
3        APRILE
4        MAGGIO
5        GIUGNO
6        LUGLIO
7        AGOSTO
8     SETTEMBRE
9       OTTOBRE
10     NOVEMBRE
11     DICEMBRE
12      GENNAIO
13     FEBBRAIO
14        MARZO
15       APRILE
16       MAGGIO
17       GIUGNO
18       LUGLIO
19       AGOSTO
20    SETTEMBRE
21      OTTOBRE
22     NOVEMBRE
23     DICEMBRE
24      GENNAIO
25     FEBBRAIO
26        MARZO
27       APRILE
28       MAGGIO
29       GIUGNO
30       LUGLIO
31       AGOSTO
32    SETTEMBRE
33      OTTOBRE
34     NOVEMBRE
35     DICEMBRE
36      GENNAIO
37     FEBBRAIO
38        MARZO
39       APRILE
40       MAGGIO
41       GIUGNO
42       LUGLIO
43       AGOSTO
44    SETTEMBRE
45      OTTOBRE
46     NOVEMBRE
47     DICEMBRE
48      GENNAIO
49     FEBBRAIO
50        MARZO
51       APRILE
52       MAGGIO
53       GIUGNO
54       LUGLIO
55       AGOSTO
56    SETTEMBRE
57      OTTOBRE
58     NOVEMBRE
59     DICEMBRE
Name: Mese, dtype: object

In [222]:
a.Guadagni.cumsum()

0       2382
1       5055
2      10011
3      14889
4      17788
5      21613
6      22639
7      24261
8      25839
9      29582
10     30520
11     32293
12     34578
13     35478
14     36674
15     39515
16     42542
17     45059
18     46039
19     49264
20     51189
21     54062
22     58485
23     60708
24     63438
25     66145
26     70673
27     75603
28     77736
29     81696
30     83412
31     87189
32     88148
33     92090
34     95155
35     97044
36     99876
37    103437
38    107513
39    108591
40    110776
41    112063
42    116408
43    118941
44    122851
45    126891
46    130177
47    133025
48    137992
49    142266
50    147035
51    148470
52    150210
53    152567
54    156964
55    160433
56    162712
57    167691
58    171214
59    174715
Name: Guadagni, dtype: int32

In [210]:
anni = a.Anno.unique()
for anno in anni:
    print("Anno:",anno)
    filtro_anno = a.Anno == anno
    somma_cumulativa_annuale = a.loc[filtro_anno, "Guadagni"].cumsum()
    print(somma_cumulativa_annuale)
    print("-"*10)


Anno: 0
0       849
1      3717
2      7077
3     11386
4     12463
5     13519
6     15855
7     19091
8     21111
9     23866
10    24851
11    26025
Name: Guadagni, dtype: int32
----------
Anno: 1
12     1230
13     6137
14    10115
15    14997
16    18318
17    20948
18    25748
19    29059
20    31826
21    33502
22    38157
23    39277
Name: Guadagni, dtype: int32
----------
Anno: 2
24     4632
25     6474
26     8152
27    10896
28    15081
29    17392
30    18882
31    23084
32    25824
33    29240
34    34047
35    36084
Name: Guadagni, dtype: int32
----------
Anno: 3
36     3882
37     4853
38     9136
39    10597
40    14398
41    18490
42    22294
43    23773
44    24714
45    27806
46    28716
47    32036
Name: Guadagni, dtype: int32
----------
Anno: 4
48     2200
49     5355
50    10218
51    14202
52    17561
53    18473
54    19794
55    22256
56    24498
57    28177
58    29782
59    30812
Name: Guadagni, dtype: int32
----------


# DIMCOSTUMER

In [32]:
query = "SELECT * FROM dimcustomer"
customer_df = pd.read_sql(sql=query, con=db_engine)
customer_df

CustomerKey  GeographyKey CustomerAlternateKey Title  FirstName  \
0            11000            26           AW00011000  None        Jon   
1            11001            37           AW00011001  None     Eugene   
2            11002            31           AW00011002  None      Ruben   
3            11003            11           AW00011003  None    Christy   
4            11004            19           AW00011004  None  Elizabeth   
...            ...           ...                  ...   ...        ...   
18479        29479           209           AW00029479  None      Tommy   
18480        29480           248           AW00029480  None       Nina   
18481        29481           120           AW00029481  None       Ivan   
18482        29482           179           AW00029482  None    Clayton   
18483        29483           217           AW00029483  None      Jésus   

      MiddleName LastName  NameStyle   BirthDate MaritalStatus  ...  \
0              V     Yang          0  1971-10-06             M  ...   
1              L    Huang          0  1976-05-10             S  ...   
2           None   Torres          0  1971-02-09             M  ...   
3           None      Zhu          0  1973-08-14             S  ...   
4           None  Johnson          0  1979-08-05             S  ...   
...          ...      ...        ...         ...           ...  ...   
18479          L     Tang          0  1969-06-30             M  ...   
18480          W     Raji          0  1977-05-06             S  ...   
18481       None     Suri          0  1965-07-04             S  ...   
18482       None    Zhang          0  1964-09-01             M  ...   
18483          L  Navarro          0  1965-06-06             M  ...   

      EnglishOccupation SpanishOccupation FrenchOccupation  HouseOwnerFlag  \
0          Professional       Profesional            Cadre               1   
1          Professional       Profesional            Cadre               0   
2          Professional       Profesional            Cadre               1   
3          Professional       Profesional            Cadre               0   
4          Professional       Profesional            Cadre               1   
...                 ...               ...              ...             ...   
18479          Clerical    Administrativo          Employé               1   
18480          Clerical    Administrativo          Employé               1   
18481          Clerical    Administrativo          Employé               0   
18482          Clerical    Administrativo          Employé               1   
18483          Clerical    Administrativo          Employé               1   

       NumberCarsOwned               AddressLine1 AddressLine2  \
0                    0            3761 N. 14th St         None   
1                    1                 2243 W St.         None   
2                    1           5844 Linden Land         None   
3                    1           1825 Village Pl.         None   
4                    4        7553 Harness Circle         None   
...                ...                        ...          ...   
18479                0          111, rue Maillard         None   
18480                0          9 Katherine Drive         None   
18481                0                Knaackstr 4         None   
18482                0     1080, quai de Grenelle         None   
18483                0  244, rue de la Centenaire         None   

                     Phone DateFirstPurchase CommuteDistance  
0      1 (11) 500 555-0162        2011-01-19       1-2 Miles  
1      1 (11) 500 555-0110        2011-01-15       0-1 Miles  
2      1 (11) 500 555-0184        2011-01-07       2-5 Miles  
3      1 (11) 500 555-0162        2010-12-29      5-10 Miles  
4      1 (11) 500 555-0131        2011-01-23       1-2 Miles  
...                    ...               ...             ...  
18479  1 (11) 500 555-0136        2012-09-04       0-1 Miles  
18480  1 (11) 500 555-0146     

In [34]:
customer_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18484 entries, 0 to 18483
Data columns (total 29 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   CustomerKey           18484 non-null  int64  
 1   GeographyKey          18484 non-null  int64  
 2   CustomerAlternateKey  18484 non-null  object 
 3   Title                 101 non-null    object 
 4   FirstName             18484 non-null  object 
 5   MiddleName            10654 non-null  object 
 6   LastName              18484 non-null  object 
 7   NameStyle             18484 non-null  int64  
 8   BirthDate             18484 non-null  object 
 9   MaritalStatus         18484 non-null  object 
 10  Suffix                3 non-null      object 
 11  Gender                18484 non-null  object 
 12  EmailAddress          18484 non-null  object 
 13  YearlyIncome          18484 non-null  float64
 14  TotalChildren         18484 non-null  int64  
 15  NumberChildrenAtHom

In [36]:
customer_df.FirstName = customer_df.FirstName.astype(str)
customer_df.MiddleName = customer_df.MiddleName.astype(str)
customer_df.LastName = customer_df.LastName.astype(str)

In [38]:
customer_df.FirstName = customer_df["FirstName"].str.lower()
customer_df.MiddleName = customer_df["MiddleName"].str.lower()
customer_df.LastName = customer_df["LastName"].str.upper()

In [40]:
customer_df

CustomerKey  GeographyKey CustomerAlternateKey Title  FirstName  \
0            11000            26           AW00011000  None        jon   
1            11001            37           AW00011001  None     eugene   
2            11002            31           AW00011002  None      ruben   
3            11003            11           AW00011003  None    christy   
4            11004            19           AW00011004  None  elizabeth   
...            ...           ...                  ...   ...        ...   
18479        29479           209           AW00029479  None      tommy   
18480        29480           248           AW00029480  None       nina   
18481        29481           120           AW00029481  None       ivan   
18482        29482           179           AW00029482  None    clayton   
18483        29483           217           AW00029483  None      jésus   

      MiddleName LastName  NameStyle   BirthDate MaritalStatus  ...  \
0              v     YANG          0  1971-10-06             M  ...   
1              l    HUANG          0  1976-05-10             S  ...   
2           none   TORRES          0  1971-02-09             M  ...   
3           none      ZHU          0  1973-08-14             S  ...   
4           none  JOHNSON          0  1979-08-05             S  ...   
...          ...      ...        ...         ...           ...  ...   
18479          l     TANG          0  1969-06-30             M  ...   
18480          w     RAJI          0  1977-05-06             S  ...   
18481       none     SURI          0  1965-07-04             S  ...   
18482       none    ZHANG          0  1964-09-01             M  ...   
18483          l  NAVARRO          0  1965-06-06             M  ...   

      EnglishOccupation SpanishOccupation FrenchOccupation  HouseOwnerFlag  \
0          Professional       Profesional            Cadre               1   
1          Professional       Profesional            Cadre               0   
2          Professional       Profesional            Cadre               1   
3          Professional       Profesional            Cadre               0   
4          Professional       Profesional            Cadre               1   
...                 ...               ...              ...             ...   
18479          Clerical    Administrativo          Employé               1   
18480          Clerical    Administrativo          Employé               1   
18481          Clerical    Administrativo          Employé               0   
18482          Clerical    Administrativo          Employé               1   
18483          Clerical    Administrativo          Employé               1   

       NumberCarsOwned               AddressLine1 AddressLine2  \
0                    0            3761 N. 14th St         None   
1                    1                 2243 W St.         None   
2                    1           5844 Linden Land         None   
3                    1           1825 Village Pl.         None   
4                    4        7553 Harness Circle         None   
...                ...                        ...          ...   
18479                0          111, rue Maillard         None   
18480                0          9 Katherine Drive         None   
18481                0                Knaackstr 4         None   
18482                0     1080, quai de Grenelle         None   
18483                0  244, rue de la Centenaire         None   

                     Phone DateFirstPurchase CommuteDistance  
0      1 (11) 500 555-0162        2011-01-19       1-2 Miles  
1      1 (11) 500 555-0110        2011-01-15       0-1 Miles  
2      1 (11) 500 555-0184        2011-01-07       2-5 Miles  
3      1 (11) 500 555-0162        2010-12-29      5-10 Miles  
4      1 (11) 500 555-0131        2011-01-23       1-2 Miles  
...                    ...               ...             ...  
18479  1 (11) 500 555-0136        2012-09-04       0-1 Miles  
18480  1 (11) 500 555-0146     

In [42]:
customer_df.columns

Index(['CustomerKey', 'GeographyKey', 'CustomerAlternateKey', 'Title',
       'FirstName', 'MiddleName', 'LastName', 'NameStyle', 'BirthDate',
       'MaritalStatus', 'Suffix', 'Gender', 'EmailAddress', 'YearlyIncome',
       'TotalChildren', 'NumberChildrenAtHome', 'EnglishEducation',
       'SpanishEducation', 'FrenchEducation', 'EnglishOccupation',
       'SpanishOccupation', 'FrenchOccupation', 'HouseOwnerFlag',
       'NumberCarsOwned', 'AddressLine1', 'AddressLine2', 'Phone',
       'DateFirstPurchase', 'CommuteDistance'],
      dtype='object')

In [44]:
customer_df.EmailAddress.str.split(pat="@")

0             [jon24, adventure-works.com]
1          [eugene10, adventure-works.com]
2           [ruben35, adventure-works.com]
3         [christy12, adventure-works.com]
4        [elizabeth5, adventure-works.com]
                       ...                
18479        [tommy2, adventure-works.com]
18480        [nina21, adventure-works.com]
18481         [ivan0, adventure-works.com]
18482      [clayton0, adventure-works.com]
18483        [jésus9, adventure-works.com]
Name: EmailAddress, Length: 18484, dtype: object

In [46]:
customer_df.EmailAddress

0             jon24@adventure-works.com
1          eugene10@adventure-works.com
2           ruben35@adventure-works.com
3         christy12@adventure-works.com
4        elizabeth5@adventure-works.com
                      ...              
18479        tommy2@adventure-works.com
18480        nina21@adventure-works.com
18481         ivan0@adventure-works.com
18482      clayton0@adventure-works.com
18483        jésus9@adventure-works.com
Name: EmailAddress, Length: 18484, dtype: object

In [48]:
customer_df.Phone.str.split()

0        [1, (11), 500, 555-0162]
1        [1, (11), 500, 555-0110]
2        [1, (11), 500, 555-0184]
3        [1, (11), 500, 555-0162]
4        [1, (11), 500, 555-0131]
                   ...           
18479    [1, (11), 500, 555-0136]
18480    [1, (11), 500, 555-0146]
18481    [1, (11), 500, 555-0144]
18482    [1, (11), 500, 555-0137]
18483    [1, (11), 500, 555-0141]
Name: Phone, Length: 18484, dtype: object

In [50]:
customer_df.Phone

0        1 (11) 500 555-0162
1        1 (11) 500 555-0110
2        1 (11) 500 555-0184
3        1 (11) 500 555-0162
4        1 (11) 500 555-0131
                ...         
18479    1 (11) 500 555-0136
18480    1 (11) 500 555-0146
18481    1 (11) 500 555-0144
18482    1 (11) 500 555-0137
18483    1 (11) 500 555-0141
Name: Phone, Length: 18484, dtype: object

In [52]:
filtro_mail_21 = customer_df.EmailAddress.str.contains("21", regex = False)

In [54]:
customer_df.loc[filtro_mail_21, "EmailAddress"]

64          chase21@adventure-works.com
118         alvin21@adventure-works.com
130      caroline21@adventure-works.com
163       gabriel21@adventure-works.com
203          luis21@adventure-works.com
                      ...              
18337      hector21@adventure-works.com
18365      cheryl21@adventure-works.com
18435     stanley21@adventure-works.com
18440      rafael21@adventure-works.com
18480        nina21@adventure-works.com
Name: EmailAddress, Length: 431, dtype: object

In [56]:
filtro_mail_20 = customer_df.EmailAddress.str.contains("20", regex = False)
filtro_mail_10 = customer_df.EmailAddress.str.contains("10", regex = False)
filtro_mail_10_20 = filtro_mail_10 | filtro_mail_20
customer_df.loc[filtro_mail_10_20,"EmailAddress"]

1            eugene10@adventure-works.com
10        jacquelyn20@adventure-works.com
22            ethan20@adventure-works.com
30          bethany10@adventure-works.com
32           denise10@adventure-works.com
                       ...               
18395         lydia10@adventure-works.com
18397        walter10@adventure-works.com
18415         oscar20@adventure-works.com
18418       lindsey10@adventure-works.com
18428    kristopher20@adventure-works.com
Name: EmailAddress, Length: 1042, dtype: object

In [58]:
#Metodo apply () per calcolare la lunghezza delle colonne
customer_df["EmailLength"] = customer_df.EmailAddress.apply(len)
customer_df

CustomerKey  GeographyKey CustomerAlternateKey Title  FirstName  \
0            11000            26           AW00011000  None        jon   
1            11001            37           AW00011001  None     eugene   
2            11002            31           AW00011002  None      ruben   
3            11003            11           AW00011003  None    christy   
4            11004            19           AW00011004  None  elizabeth   
...            ...           ...                  ...   ...        ...   
18479        29479           209           AW00029479  None      tommy   
18480        29480           248           AW00029480  None       nina   
18481        29481           120           AW00029481  None       ivan   
18482        29482           179           AW00029482  None    clayton   
18483        29483           217           AW00029483  None      jésus   

      MiddleName LastName  NameStyle   BirthDate MaritalStatus  ...  \
0              v     YANG          0  1971-10-06             M  ...   
1              l    HUANG          0  1976-05-10             S  ...   
2           none   TORRES          0  1971-02-09             M  ...   
3           none      ZHU          0  1973-08-14             S  ...   
4           none  JOHNSON          0  1979-08-05             S  ...   
...          ...      ...        ...         ...           ...  ...   
18479          l     TANG          0  1969-06-30             M  ...   
18480          w     RAJI          0  1977-05-06             S  ...   
18481       none     SURI          0  1965-07-04             S  ...   
18482       none    ZHANG          0  1964-09-01             M  ...   
18483          l  NAVARRO          0  1965-06-06             M  ...   

      SpanishOccupation FrenchOccupation HouseOwnerFlag  NumberCarsOwned  \
0           Profesional            Cadre              1                0   
1           Profesional            Cadre              0                1   
2           Profesional            Cadre              1                1   
3           Profesional            Cadre              0                1   
4           Profesional            Cadre              1                4   
...                 ...              ...            ...              ...   
18479    Administrativo          Employé              1                0   
18480    Administrativo          Employé              1                0   
18481    Administrativo          Employé              0                0   
18482    Administrativo          Employé              1                0   
18483    Administrativo          Employé              1                0   

                    AddressLine1  AddressLine2                Phone  \
0                3761 N. 14th St          None  1 (11) 500 555-0162   
1                     2243 W St.          None  1 (11) 500 555-0110   
2               5844 Linden Land          None  1 (11) 500 555-0184   
3               1825 Village Pl.          None  1 (11) 500 555-0162   
4            7553 Harness Circle          None  1 (11) 500 555-0131   
...                          ...           ...                  ...   
18479          111, rue Maillard          None  1 (11) 500 555-0136   
18480          9 Katherine Drive          None  1 (11) 500 555-0146   
18481                Knaackstr 4          None  1 (11) 500 555-0144   
18482     1080, quai de Grenelle          None  1 (11) 500 555-0137   
18483  244, rue de la Centenaire          None  1 (11) 500 555-0141   

      DateFirstPurchase CommuteDistance EmailLength  
0            2011-01-19       1-2 Miles          25  
1            2011-01-15       0-1 Miles          28  
2            2011-01-07       2-5 Miles          27  
3            2010-12-29      5-10 Miles          29  
4            2011-01-23       1-2 Miles          30  
...                 ...             ...         ...  
18479        2012-09-04       0-1 Miles          26  
18480        2013-07-17       0-1 Miles          26  
18481        

In [60]:
customer_mail_sorted = customer_df.sort_values("EmailLength")
customer_mail_sorted.head(10).EmailAddress

3974       k0@adventure-works.com
15299      y0@adventure-works.com
1658     joy2@adventure-works.com
550      deb4@adventure-works.com
13888    roy9@adventure-works.com
6264     ian1@adventure-works.com
2326     bob8@adventure-works.com
13825    zoe3@adventure-works.com
215      ana7@adventure-works.com
11213    jay9@adventure-works.com
Name: EmailAddress, dtype: object

In [62]:
customer_mail_sorted.tail(10).EmailAddress

4772     christopher10@adventure-works.com
9242     christopher11@adventure-works.com
1767     christopher18@adventure-works.com
16808    christopher19@adventure-works.com
8768     christopher12@adventure-works.com
3512     christopher27@adventure-works.com
4973     christopher16@adventure-works.com
13417    christopher14@adventure-works.com
2655     christopher23@adventure-works.com
16661    christopher17@adventure-works.com
Name: EmailAddress, dtype: object

In [64]:
customer_df.EmailAddress.str.replace("adventure-works.com","aw-db.com")

0             jon24@aw-db.com
1          eugene10@aw-db.com
2           ruben35@aw-db.com
3         christy12@aw-db.com
4        elizabeth5@aw-db.com
                 ...         
18479        tommy2@aw-db.com
18480        nina21@aw-db.com
18481         ivan0@aw-db.com
18482      clayton0@aw-db.com
18483        jésus9@aw-db.com
Name: EmailAddress, Length: 18484, dtype: object

In [66]:
street = customer_df.AddressLine1.str.contains("Street", regex=False)
customer_df.loc[street, "AddressLine1"]

7              942 Brook Street
12            4785 Scott Street
32           626 Bentley Street
63           4927 Virgil Street
125          8481 Zartop Street
                  ...          
18273        2155 Zartop Street
18292         3931 Scott Street
18323    134 Peachwillow Street
18333       1023 Hawkins Street
18351          8618 Rose Street
Name: AddressLine1, Length: 512, dtype: object

# FACEBOOK

In [119]:
file_path = "C:/Users/HP/Downloads/archive/beginner_datasets/facebook.csv"
fb = pd.read_csv(file_path)
fb

status_id status_type status_published  \
0      246675545449582_1649696485147474       video   4/22/2018 6:00   
1      246675545449582_1649426988507757       photo  4/21/2018 22:45   
2      246675545449582_1648730588577397       video   4/21/2018 6:17   
3      246675545449582_1648576705259452       photo   4/21/2018 2:29   
4      246675545449582_1645700502213739       photo   4/18/2018 3:22   
...                                 ...         ...              ...   
7045  1050855161656896_1061863470556065       photo   9/24/2016 2:58   
7046  1050855161656896_1061334757275603       photo  9/23/2016 11:19   
7047  1050855161656896_1060126464063099       photo  9/21/2016 23:03   
7048  1050855161656896_1058663487542730       photo   9/20/2016 0:43   
7049  1050855161656896_1050858841656528       photo  9/10/2016 10:30   

      num_reactions  num_comments  num_shares  num_likes  num_loves  num_wows  \
0               529           512         262        432         92         3   
1               150             0           0        150          0         0   
2               227           236          57        204         21         1   
3               111             0           0        111          0         0   
4               213             0           0        204          9         0   
...             ...           ...         ...        ...        ...       ...   
7045             89             0           0         89          0         0   
7046             16             0           0         14          1         0   
7047              2             0           0          1          1         0   
7048            351            12          22        349          2         0   
7049             17             0           0         17          0         0   

      num_hahas  num_sads  num_angrys  
0             1         1           0  
1             0         0           0  
2             1         0           0  
3             0         0           0  
4             0         0           0  
...         ...       ...         ...  
7045          0         0           0  
7046          1         0           0  
7047          0         0           0  
7048          0         0           0  
7049          0         0           0  

[7050 rows x 12 columns]

In [121]:
fb.status_published = pd.to_datetime(fb.status_published, format="%m/%d/%Y %H:%M")

In [123]:
fb.head(10)

status_id status_type    status_published  \
0  246675545449582_1649696485147474       video 2018-04-22 06:00:00   
1  246675545449582_1649426988507757       photo 2018-04-21 22:45:00   
2  246675545449582_1648730588577397       video 2018-04-21 06:17:00   
3  246675545449582_1648576705259452       photo 2018-04-21 02:29:00   
4  246675545449582_1645700502213739       photo 2018-04-18 03:22:00   
5  246675545449582_1645650162218773       photo 2018-04-18 02:14:00   
6  246675545449582_1645564175560705       video 2018-04-18 00:24:00   
7  246675545449582_1644824665634656       video 2018-04-17 07:42:00   
8  246675545449582_1644655795651543       photo 2018-04-17 03:33:00   
9  246675545449582_1638788379571618       photo 2018-04-11 04:53:00   

   num_reactions  num_comments  num_shares  num_likes  num_loves  num_wows  \
0            529           512         262        432         92         3   
1            150             0           0        150          0         0   
2            227           236          57        204         21         1   
3            111             0           0        111          0         0   
4            213             0           0        204          9         0   
5            217             6           0        211          5         1   
6            503           614          72        418         70        10   
7            295           453          53        260         32         1   
8            203             1           0        198          5         0   
9            170             9           1        167          3         0   

   num_hahas  num_sads  num_angrys  
0          1         1           0  
1          0         0           0  
2          1         0           0  
3          0         0           0  
4          0         0           0  
5          0         0           0  
6          2         0           3  
7          1         0           1  
8          0         0           0  
9          0         0           0

In [125]:
fb.status_published.dt.strftime("%Y-%m-%d %H:%M:%S")

0       2018-04-22 06:00:00
1       2018-04-21 22:45:00
2       2018-04-21 06:17:00
3       2018-04-21 02:29:00
4       2018-04-18 03:22:00
               ...         
7045    2016-09-24 02:58:00
7046    2016-09-23 11:19:00
7047    2016-09-21 23:03:00
7048    2016-09-20 00:43:00
7049    2016-09-10 10:30:00
Name: status_published, Length: 7050, dtype: object

In [127]:
fb.status_published.dt.dayofweek

0       6
1       5
2       5
3       5
4       2
       ..
7045    5
7046    4
7047    2
7048    1
7049    5
Name: status_published, Length: 7050, dtype: int32

In [129]:
post_2012 = fb.status_published.dt.year == 2012

In [131]:
fb.loc[post_2012]

status_id status_type    status_published  \
2046  246675545449582_303160673134402       photo 2012-12-23 03:11:00   
2047  246675545449582_302341609882975       photo 2012-12-21 06:02:00   
2048  246675545449582_299212316862571       photo 2012-12-13 11:36:00   
2049  246675545449582_290159194434550       photo 2012-11-18 09:50:00   
2050  246675545449582_290154177768385       photo 2012-11-18 09:16:00   
...                               ...         ...                 ...   
2630  246675545449582_246688625448274       photo 2012-07-15 03:54:00   
2631  246675545449582_246686478781822       photo 2012-07-15 03:42:00   
2632  246675545449582_246684432115360       photo 2012-07-15 03:32:00   
2633  246675545449582_246678538782616       photo 2012-07-15 02:58:00   
2634  246675545449582_246677465449390       photo 2012-07-15 02:51:00   

      num_reactions  num_comments  num_shares  num_likes  num_loves  num_wows  \
2046              0             0           0          0          0         0   
2047              4             1           0          4          0         0   
2048              3             0           0          3          0         0   
2049              5             1           1          5          0         0   
2050              2             0           0          2          0         0   
...             ...           ...         ...        ...        ...       ...   
2630             19            17           3         19          0         0   
2631             12             3           0         12          0         0   
2632             14             1           0         14          0         0   
2633             14             7           0         14          0         0   
2634             15             3           0         15          0         0   

      num_hahas  num_sads  num_angrys  
2046          0         0           0  
2047          0         0           0  
2048          0         0           0  
2049          0         0           0  
2050          0         0           0  
...         ...       ...         ...  
2630          0         0           0  
2631          0         0           0  
2632          0         0           0  
2633          0         0           0  
2634          0         0           0  

[589 rows x 12 columns]

In [82]:
post_2018 = fb.status_published.dt.year == 2018
post_may = fb.status_published.dt.month == 5
post_may_2018 = post_2018 & post_may
fb.loc[post_may_2018]

status_id status_type    status_published  \
2690    134115277150304_264554297439734       photo 2018-05-31 12:17:00   
2691    134115277150304_264493694112461       photo 2018-05-31 08:37:00   
2692    134115277150304_264416247453539       photo 2018-05-31 03:12:00   
2693    134115277150304_264168277478336       photo 2018-05-30 08:46:00   
2694    134115277150304_264160757479088       photo 2018-05-30 08:24:00   
...                                 ...         ...                 ...   
6339  1050855161656896_1615666458509094        link 2018-05-04 01:27:00   
6340  1050855161656896_1614717551937318       photo 2018-05-03 01:15:00   
6341  1050855161656896_1614187798656960       photo 2018-05-02 12:17:00   
6342  1050855161656896_1613063202102753       video 2018-05-01 08:38:00   
6343  1050855161656896_1612778858797854       photo 2018-05-01 01:18:00   

      num_reactions  num_comments  num_shares  num_likes  num_loves  num_wows  \
2690             11             0           0         10          0         1   
2691             56             4           2         55          1         0   
2692             72            13           1         69          3         0   
2693             10             0           0         10          0         0   
2694             56             3          12         54          2         0   
...             ...           ...         ...        ...        ...       ...   
6339             12             7           7         12          0         0   
6340             63            21           1         62          1         0   
6341            218           105         153        212          2         4   
6342            246          2373         475        108        130         5   
6343             34             5           0         33          1         0   

      num_hahas  num_sads  num_angrys  
2690          0         0           0  
2691          0         0           0  
2692          0         0           0  
2693          0         0           0  
2694          0         0           0  
...         ...       ...         ...  
6339          0         0           0  
6340          0         0           0  
6341          0         0           0  
6342          2         1           0  
6343          0         0           0  

[595 rows x 12 columns]

In [133]:
monday = fb.status_published.dt.dayofweek == 0
tuesday = fb.status_published.dt.dayofweek == 1
wednesday = fb.status_published.dt.dayofweek == 2
thursday = fb.status_published.dt.dayofweek == 3
friday = fb.status_published.dt.dayofweek == 4
saturday = fb.status_published.dt.dayofweek == 5
sunday = fb.status_published.dt.dayofweek == 6
weekend = saturday | sunday
not_weekend = monday | tuesday | wednesday | thursday | friday
post_weekend = fb.loc[weekend]
post_not_weekend = fb.loc[not_weekend]

In [135]:
post_weekend

status_id status_type    status_published  \
0      246675545449582_1649696485147474       video 2018-04-22 06:00:00   
1      246675545449582_1649426988507757       photo 2018-04-21 22:45:00   
2      246675545449582_1648730588577397       video 2018-04-21 06:17:00   
3      246675545449582_1648576705259452       photo 2018-04-21 02:29:00   
12     246675545449582_1635846033199186       photo 2018-04-08 05:10:00   
...                                 ...         ...                 ...   
7042  1050855161656896_1061944223881323       photo 2016-09-24 06:07:00   
7043  1050855161656896_1061918183883927       photo 2016-09-24 05:07:00   
7044  1050855161656896_1061906620551750       photo 2016-09-24 04:37:00   
7045  1050855161656896_1061863470556065       photo 2016-09-24 02:58:00   
7049  1050855161656896_1050858841656528       photo 2016-09-10 10:30:00   

      num_reactions  num_comments  num_shares  num_likes  num_loves  num_wows  \
0               529           512         262        432         92         3   
1               150             0           0        150          0         0   
2               227           236          57        204         21         1   
3               111             0           0        111          0         0   
12              313             4           2        305          6         2   
...             ...           ...         ...        ...        ...       ...   
7042              4             0           0          4          0         0   
7043            196             2           3        195          1         0   
7044             86             0           0         86          0         0   
7045             89             0           0         89          0         0   
7049             17             0           0         17          0         0   

      num_hahas  num_sads  num_angrys  
0             1         1           0  
1             0         0           0  
2             1         0           0  
3             0         0           0  
12            0         0           0  
...         ...       ...         ...  
7042          0         0           0  
7043          0         0           0  
7044          0         0           0  
7045          0         0           0  
7049          0         0           0  

[2041 rows x 12 columns]

In [137]:
post_not_weekend

status_id status_type    status_published  \
4      246675545449582_1645700502213739       photo 2018-04-18 03:22:00   
5      246675545449582_1645650162218773       photo 2018-04-18 02:14:00   
6      246675545449582_1645564175560705       video 2018-04-18 00:24:00   
7      246675545449582_1644824665634656       video 2018-04-17 07:42:00   
8      246675545449582_1644655795651543       photo 2018-04-17 03:33:00   
...                                 ...         ...                 ...   
7037  1050855161656896_1070127156396363       photo 2016-10-04 01:49:00   
7039  1050855161656896_1063592313716514       photo 2016-09-26 05:48:00   
7046  1050855161656896_1061334757275603       photo 2016-09-23 11:19:00   
7047  1050855161656896_1060126464063099       photo 2016-09-21 23:03:00   
7048  1050855161656896_1058663487542730       photo 2016-09-20 00:43:00   

      num_reactions  num_comments  num_shares  num_likes  num_loves  num_wows  \
4               213             0           0        204          9         0   
5               217             6           0        211          5         1   
6               503           614          72        418         70        10   
7               295           453          53        260         32         1   
8               203             1           0        198          5         0   
...             ...           ...         ...        ...        ...       ...   
7037             15             0           1         14          1         0   
7039              7             0           0          7          0         0   
7046             16             0           0         14          1         0   
7047              2             0           0          1          1         0   
7048            351            12          22        349          2         0   

      num_hahas  num_sads  num_angrys  
4             0         0           0  
5             0         0           0  
6             2         0           3  
7             1         0           1  
8             0         0           0  
...         ...       ...         ...  
7037          0         0           0  
7039          0         0           0  
7046          1         0           0  
7047          0         0           0  
7048          0         0           0  

[5009 rows x 12 columns]

In [139]:
fb_sorted_years = fb.sort_values("status_published")

In [141]:
fb_sorted_years.groupby(fb_sorted_years.status_published.dt.year).head(1)

status_id status_type    status_published  \
2634   246675545449582_246677465449390       photo 2012-07-15 02:51:00   
2045   246675545449582_307286656055137       photo 2013-01-02 02:23:00   
1545   246675545449582_486434424807025       video 2014-01-05 09:23:00   
1312   246675545449582_682315955218870       video 2015-01-02 03:41:00   
975    246675545449582_868621673254963       photo 2016-01-03 04:23:00   
434   246675545449582_1159059407544520       photo 2017-01-02 08:25:00   
3264   134115277150304_211295932765571       video 2018-01-01 01:39:00   

      num_reactions  num_comments  num_shares  num_likes  num_loves  num_wows  \
2634             15             3           0         15          0         0   
2045             25             3           1         25          0         0   
1545            460            31           1        460          0         0   
1312           1294            23          47       1294          0         0   
975              21             2           0         21          0         0   
434             254             5           0        252          2         0   
3264            296           869          58        252         39         2   

      num_hahas  num_sads  num_angrys  
2634          0         0           0  
2045          0         0           0  
1545          0         0           0  
1312          0         0           0  
975           0         0           0  
434           0         0           0  
3264          2         1           0

In [143]:
fb_sorted_years.groupby(fb_sorted_years.status_published.dt.year).tail(1)

status_id status_type    status_published  \
2046   246675545449582_303160673134402       photo 2012-12-23 03:11:00   
1546   246675545449582_484017401715394       video 2013-12-31 23:38:00   
1313   246675545449582_668776289906170       video 2014-12-10 07:42:00   
976    246675545449582_867129116737552       photo 2015-12-31 08:20:00   
435   246675545449582_1157109567739504       photo 2016-12-31 05:13:00   
3265   134115277150304_211240079437823       photo 2017-12-31 22:15:00   
4730   614855718638584_855501374574016       photo 2018-06-13 01:12:00   

      num_reactions  num_comments  num_shares  num_likes  num_loves  num_wows  \
2046              0             0           0          0          0         0   
1546            100             0           0        100          0         0   
1313           1501           113          38       1501          0         0   
976             318             7           1        318          0         0   
435             304             9           6        298          5         1   
3265             45             2           0         45          0         0   
4730              2             0           0          2          0         0   

      num_hahas  num_sads  num_angrys  
2046          0         0           0  
1546          0         0           0  
1313          0         0           0  
976           0         0           0  
435           0         0           0  
3265          0         0           0  
4730          0         0           0

In [145]:
fb.status_published.loc[25].tz_localize("Europe/Rome")

Timestamp('2018-03-19 22:34:00+0100', tz='Europe/Rome')

# POKEMON

In [113]:
file_path2 = "C:/Users/HP/Downloads/archive/beginner_datasets/pokemon.csv"
pokemon = pd.read_csv(file_path2)
pokemon.head(10)

#                       Name Type 1  Type 2  Total  HP  Attack  Defense  \
0  1                  Bulbasaur  Grass  Poison    318  45      49       49   
1  2                    Ivysaur  Grass  Poison    405  60      62       63   
2  3                   Venusaur  Grass  Poison    525  80      82       83   
3  3      VenusaurMega Venusaur  Grass  Poison    625  80     100      123   
4  4                 Charmander   Fire     NaN    309  39      52       43   
5  5                 Charmeleon   Fire     NaN    405  58      64       58   
6  6                  Charizard   Fire  Flying    534  78      84       78   
7  6  CharizardMega Charizard X   Fire  Dragon    634  78     130      111   
8  6  CharizardMega Charizard Y   Fire  Flying    634  78     104       78   
9  7                   Squirtle  Water     NaN    314  44      48       65   

   Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0       65       65     45           1      False  
1       80       80     60           1      False  
2      100      100     80           1      False  
3      122      120     80           1      False  
4       60       50     65           1      False  
5       80       65     80           1      False  
6      109       85    100           1      False  
7      130       85    100           1      False  
8      159      115    100           1      False  
9       50       64     43           1      False

In [157]:
pokemon.isnull().sum() #.isnull e .isna danno lo stesso risultato

#               0
Name            0
Type 1          0
Type 2        386
Total           0
HP              0
Attack          0
Defense         0
Sp. Atk         0
Sp. Def         0
Speed           0
Generation      0
Legendary       0
dtype: int64

Ci sono 386 valori nulli nel Type 2.  
Penso che non avrebbe senso cercare di riempire la casella dei tipi nulli, poiché essi
dovrebbero riguardare l'evoluzione del pokemon. (Non lo so, non mi sono mai piaciuti i pokemon).  
In ogni caso si potrebbe riempire i valori nulli con il valore 0.

In [117]:
pokemon.dropna() #Sono rimasti solamente i pokemon con l'evoluzione.

#                   Name   Type 1  Type 2  Total  HP  Attack  Defense  \
0      1              Bulbasaur    Grass  Poison    318  45      49       49   
1      2                Ivysaur    Grass  Poison    405  60      62       63   
2      3               Venusaur    Grass  Poison    525  80      82       83   
3      3  VenusaurMega Venusaur    Grass  Poison    625  80     100      123   
6      6              Charizard     Fire  Flying    534  78      84       78   
..   ...                    ...      ...     ...    ...  ..     ...      ...   
795  719                Diancie     Rock   Fairy    600  50     100      150   
796  719    DiancieMega Diancie     Rock   Fairy    700  50     160      110   
797  720    HoopaHoopa Confined  Psychic   Ghost    600  80     110       60   
798  720     HoopaHoopa Unbound  Psychic    Dark    680  80     160       60   
799  721              Volcanion     Fire   Water    600  80     110      120   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0         65       65     45           1      False  
1         80       80     60           1      False  
2        100      100     80           1      False  
3        122      120     80           1      False  
6        109       85    100           1      False  
..       ...      ...    ...         ...        ...  
795      100      150     50           6       True  
796      160      110    110           6       True  
797      150      130     70           6       True  
798      170      130     80           6       True  
799      130       90     70           6       True  

[414 rows x 13 columns]

# AUTOMOBILI
## From Beginner Dataset

In [160]:
file_path2 = "C:/Users/HP/Downloads/archive/beginner_datasets/automobile.csv"
auto = pd.read_csv(file_path2)
auto.head(10)

symboling  normalized-losses         make fuel-type aspiration  \
0          3                NaN  alfa-romero       gas        std   
1          3                NaN  alfa-romero       gas        std   
2          1                NaN  alfa-romero       gas        std   
3          2              164.0         audi       gas        std   
4          2              164.0         audi       gas        std   
5          2                NaN         audi       gas        std   
6          1              158.0         audi       gas        std   
7          1                NaN         audi       gas        std   
8          1              158.0         audi       gas      turbo   
9          2              192.0          bmw       gas        std   

  num-of-doors   body-style drive-wheels engine-location  wheel-base  ...  \
0          two  convertible          rwd           front        88.6  ...   
1          two  convertible          rwd           front        88.6  ...   
2          two    hatchback          rwd           front        94.5  ...   
3         four        sedan          fwd           front        99.8  ...   
4         four        sedan          4wd           front        99.4  ...   
5          two        sedan          fwd           front        99.8  ...   
6         four        sedan          fwd           front       105.8  ...   
7         four        wagon          fwd           front       105.8  ...   
8         four        sedan          fwd           front       105.8  ...   
9          two        sedan          rwd           front       101.2  ...   

   engine-size  fuel-system  bore  stroke compression-ratio horsepower  \
0          130         mpfi  3.47    2.68               9.0        111   
1          130         mpfi  3.47    2.68               9.0        111   
2          152         mpfi  2.68    3.47               9.0        154   
3          109         mpfi  3.19     3.4              10.0        102   
4          136         mpfi  3.19     3.4               8.0        115   
5          136         mpfi  3.19     3.4               8.5        110   
6          136         mpfi  3.19     3.4               8.5        110   
7          136         mpfi  3.19     3.4               8.5        110   
8          131         mpfi  3.13     3.4               8.3        140   
9          108         mpfi   3.5     2.8               8.8        101   

   peak-rpm city-mpg highway-mpg  price  
0      5000       21          27  13495  
1      5000       21          27  16500  
2      5000       19          26  16500  
3      5500       24          30  13950  
4      5500       18          22  17450  
5      5500       19          25  15250  
6      5500       19          25  17710  
7      5500       19          25  18920  
8      5500       17          20  23875  
9      5800       23          29  16430  

[10 rows x 26 columns]

• Ci sono valori nulli? Dove? Quanti?

In [165]:
auto.isna().sum()

symboling             0
normalized-losses    37
make                  0
fuel-type             0
aspiration            0
num-of-doors          2
body-style            0
drive-wheels          0
engine-location       0
wheel-base            0
length                0
width                 0
height                0
curb-weight           0
engine-type           0
num-of-cylinders      0
engine-size           0
fuel-system           0
bore                  0
stroke                0
compression-ratio     0
horsepower            0
peak-rpm              0
city-mpg              0
highway-mpg           0
price                 0
dtype: int64

• Quali righe hanno un valore nullo nella colonna num-of-doors?

In [172]:
auto_senza_porte = auto["num-of-doors"].isna().sum()

In [176]:
print(f"Nella colonna num-of-doors ci sono {auto_senza_porte} valori nulli.")

Nella colonna num-of-doors ci sono 2 valori nulli.


In [182]:
auto.loc[auto["num-of-doors"].isna()]

symboling  normalized-losses   make fuel-type aspiration num-of-doors  \
26          1              148.0  dodge       gas      turbo          NaN   
60          0                NaN  mazda    diesel        std          NaN   

   body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
26      sedan          fwd           front        93.7  ...           98   
60      sedan          fwd           front        98.8  ...          122   

    fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
26         mpfi  3.03    3.39               7.6        102      5500       24   
60          idi  3.39    3.39              22.7         64      4650       36   

   highway-mpg  price  
26          30   8558  
60          42  10795  

[2 rows x 26 columns]

In [188]:
os.listdir()

['.env',
 '.ipynb_checkpoints',
 'dotenv_install.ipynb',
 'esercitazione1.ipynb',
 'install.ipynb',
 'libraries1.ipynb',
 'loopex1.ipynb',
 'Materiali per la pratica 1.zip',
 'Materiali per la pratica 2.zip',
 'Materiali per la teoria.zip',
 'test1.ipynb',
 'titanic.ipynb',
 'view_factresellersales.ipynb',
 'W13D3.ipynb',
 'W14D4.ipynb',
 'W15D1.ipynb',
 'W15D4.ipynb']

In [192]:
import numpy as np, pandas as pd
temp = pd.DataFrame({"Giorno": [0, 1, 2, 3, 4,5, 6, 7, 8, 9, 10, 11, 12],
"Temperature": [18, 19, 18, np.nan, 21, 20, 20, np.nan, 21, 23, np.nan, 23, 24]})

In [194]:
temp

Giorno  Temperature
0        0         18.0
1        1         19.0
2        2         18.0
3        3          NaN
4        4         21.0
5        5         20.0
6        6         20.0
7        7          NaN
8        8         21.0
9        9         23.0
10      10          NaN
11      11         23.0
12      12         24.0

In [234]:
temp_crescente = temp.Temperature.cumsum()
temp_cumulativa = temp_crescente.loc[len(temp)-1]
media_temp = temp_cumulativa/len(temp)
#Vorrei mettere il valore di media_temp come valore riempitivo

In [232]:
temp.interpolate()

Giorno  Temperature
0        0         18.0
1        1         19.0
2        2         18.0
3        3         19.5
4        4         21.0
5        5         20.0
6        6         20.0
7        7         20.5
8        8         21.0
9        9         23.0
10      10         23.0
11      11         23.0
12      12         24.0

pandas.DataFrame.interpolate mi aiuta a riempire i valori nulli numerici.  
Se non specificato, di default i valori mancanti sono sostituiti da valori intermedi.